Make Report.getOutputLocation return a Property instead of a Provider (#22232)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc

Previously, you had to explicitly configure PMD to ignore default rules with `ruleSets = []`.
In the Gradle 8.0, setting `ruleSetConfig` or `ruleSetFiles` to a non-empty value implicitly ignores default rules.

==== Changing test framework after setting test framework options is now an error

When configuring the built-in test task for Java, Groovy, and Scala projects, Gradle no longer allows you to
change the test framework used by the `Test` task after configuring options. This was deprecated since it silently
discarded configuration in some cases.

The following code example now produces an error:
```
test {
   options {
   }

   useJUnitPlatform()
}
```
Instead, you can:

- set the test framework before configuring options
- migrate to the <<jvm_test_suite_plugin#jvm_test_suite_plugin,JVM Test Suite Plugin>>
```
test {
   // select test framework before configuring options
   useJUnitPlatform()
   options {
   }
}
```

> ==== Report `getOutputLocation` return type changed

Reviewer: ```suggestion
==== Report `getOutputLocation` return type changed from Provider to Property
```
Suggestion: make the section title more descriptive, provide a hint to readers since non-native English speakers especially might have trouble seeing this change inline at first glance

> In the link:{groovyDslPath}/org.gradle.api.reporting.Report.html#org.gradle.api.reporting.Report[Report] interface the return type of the `outputLocation` property is now `Property<? extends FileSystemLocation>`.  It was previously `Provider<? extends FileSystemLocation>`.

Reviewer: ```suggestion
The `outputLocation` property of the link:{groovyDslPath}/org.gradle.api.reporting.Report.html#org.gradle.api.reporting.Report[Report] now returns a value of type `Property<? extends FileSystemLocation>`.  Previously, `outputLocation` returned a value of type `Provider<? extends FileSystemLocation>`.
```
Suggestion: Switch to active phrasing. Use parallel sentence construction when possible to make paragraphs more scannable.

> This change makes the Report API more internally consistent, and allows for more idiomatic configuration of reporting tasks, using:

Reviewer: ```suggestion
This change makes the Report API more internally consistent, and allows for more idiomatic configuration of reporting tasks with the following syntax:
```

```groovy
tasks.named('test') {
    reports.junitXml.outputLocation = layout.buildDirectory.dir('reports/my-report')
}
```

versus the now `@Deprecated` former usage of:

```groovy
tasks.named('test') {
    reports.junitXml.setDestination(layout.buildDirectory.file('reports/my-report-old').get().asFile)
}

> ```

Reviewer: Consider putting the old version of the usage first or eliminating it entirely. Putting the deprecated usage after the new usage can be confusing for readers, who often expect old *before* new.

Author: Good idea.

> This interface is implemented by many built-in and custom reports such as those used by JUnit.

Reviewer: ```suggestion
Many built-in and custom reports, such as those used by JUnit, implement this interface.
```
Suggestion: switch to active phrasing

> Plugins that compiled against an earlier version of Gradle containing the previous method signature may need to be recompiled to be used with newer versions of Gradle containing the new signature.

Reviewer: ```suggestion
Plugins compiled against an earlier version of Gradle containing the previous method signature may need to be recompiled to be used with newer versions of Gradle containing the new signature.
```
